### PR TITLE
Collect student feedback

### PR DESCRIPTION
**Pull request by rgu0114**
### Summary 

- provide a means for students to give feedback on the TA who just finished helping them
- prompt students with a feedback modal that pops up after the TA marks their question as done
- enable both numeric star feedback and qualitative feedback the student writes
- create Firebase functions to save this feedback object into the course document
- feedback object stores OH session, question ID, rating, and written feedback

Demo of what the entire process looks like, with the prompt appearing properly and the student's input appearing in Firestore

https://github.com/cornell-dti/office-hours/assets/100158738/47afc68d-0a77-4e56-9961-09826e7849cf

Below, we have a rudimentary version of what the feedback modal will look like

Screen Shot 2023-11-15 at 12 59 45 AM

### Documentation
included in this PR as specification comments

### Test Plan 
- open two terminals, one as a TA and the other as a student, and see that the feedback modal properly appears when question is marked as done
- check that the feedback data properly appears on the Firebase console with the associated course

### Notes 

The styling in this PR is definitely a work in progress. Substitutions for the star rating in MUI's icon library that look more similar to the design or just larger in general weren't found, but will keep looking for other ways.

Will need to discuss form validation – e.g. whether or not we're able to submit incomplete feedback and the associated warnings that may pop up if we don't have sufficient feedback given.

Also note that this is a New PR for [this ticket](https://www.notion.so/Star-rating-after-each-question-space-for-feedback-50f53284f184457badf24b5a72202de6) without all the annoying linting issues from [this PR](https://github.com/cornell-dti/office-hours/pull/761) 

### Breaking Changes 

None

- [ ] I have updated the documentation accordingly.
- [ ] My PR adds a @ts-ignore

### CODE REVIEW

**Review by Sydney-tran — COMMENTED** [review]
This PR looks really good so far! I think the feedback form will be very helpful for professors/tas. In terms of validation, I think students should be required to add a somewhat detailed explanation (one or two words shouldn't be enough) so maybe the code could enforce a minimum character/word count, but this can definitely be discussed more with everyone at standup. Also, once you finish up the styling and the PR as a whole, don't forget to remove the `console.log` statements. Besides that, very nice job!

**Review by burninglilies — CHANGES_REQUESTED** [review]
Great work so far! I don't see any bugs on my end when I run and test it for edge cases as well so far so the front end works perfect. But one thing to note is that the feedback is stored in courses and only the feedback rating and written feedback is stored. Sometimes there are multiple TA's that are assigned to a an office hours session and multiple sessions in a course so I think it would be beneficial to store the data on the student making the feedback and the TA receiving the feedback. And then maybe storing this in sessions instead of courses or as an alternative to this, store the session of the feedback in courses with the other information

**Review by stevenyuser — COMMENTED** [review]
This feature looks really good! I think the design looks pretty good and I tested it on my machine with two different accounts and it works. I also think it's a pretty useful feature for TAs, Professors, and students.

I have three possible suggestions about changing the functionality of the "Submit Rating" button. 
1. The button is clickable even when the conditions for valid feedback aren't fulfilled. Although this is visually shown to the user through the use of a grayed-out color for the button, the button isn't actually disabled. I recommend adding a `disabled` property to the button to fix this.
2. The logic for valid feedback is repeated and should probably be in its own function or state variable. This way, it will be easier to call the logic for if feedback is valid and it will also be easier to add more complex logic to see if feedback is valid.
3. Currently, feedback is only accepted if there is a rating and there's some text in the feedback box (length > 0). However, some students will probably only give a star rating to a TA and not write anything. I think that feedback should also be accepted, as it will probably be one of the most common forms of feedback.

File: src/components/includes/FeedbackPrompt.tsx [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                    variant="contained"
+                    onClick={() => {
+                        if (rating && feedback?.length > 0) {
+                            props.onClose(rating || undefined, feedback); 
+                        }
+                        props.closeFeedbackPrompt();
+                    }}
+                    style={{position: "absolute", bottom: "2rem", right: "1.8rem"}}
+                    color={(rating && feedback?.length > 0) ? "primary" : "default"}
+
```
[the remark this answers is not in the document]
Comment (rgu0114) [reply src/components/includes/FeedbackPrompt.tsx:97]: Thanks for the feedback! These are definitely great points – I just pushed the new changes. I added the disabled property and only made the existence of `rating` a prerequisite for submitting feedback.

**Review by burninglilies — APPROVED** [review]
Thanks for making the changes Richard! I remember Alisha and Benji were the ones who suggested making comments on a review required to help collect extra feedback as justification to a star review. But we can leave it optional for now because of convenience purposes. 
Also, I tested for the same things as my original review and it looks good as well.
